services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,428 @@
+
+        return MongoDbDeviceRegistryUtils.isModificationEnabled(config)
+                .compose(ok -> processDeleteTenant(tenantId, resourceVersion, span))
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<Result<Void>> processDeleteTenant(final String tenantId, final Optional<String> resourceVersion,
+            final Span span) {
+
+        final JsonObject deleteTenantQuery = MongoDbDocumentBuilder.forVersion(resourceVersion)
+                .withTenantId(tenantId)
+                .document();
+
+        final Promise<JsonObject> deleteTenantPromise = Promise.promise();
+        mongoClient.findOneAndDelete(config.getCollectionName(), deleteTenantQuery, deleteTenantPromise);
+        return deleteTenantPromise.future()
+                .compose(tenantDtoResult -> Optional.ofNullable(tenantDtoResult)
+                        .map(deleted -> {
+                            span.log("successfully deleted tenant");
+                            return Future.succeededFuture(Result.<Void> from(HttpURLConnection.HTTP_NO_CONTENT));
+                        })
+                        .orElseGet(() -> MongoDbDeviceRegistryUtils.checkForVersionMismatchAndFail(tenantId,
+                                resourceVersion, findTenant(tenantId))));
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final String tenantId) {
+        Objects.requireNonNull(tenantId);
+
+        return get(tenantId, NoopSpan.INSTANCE);
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final String tenantId, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(span);
+
+        return readTenant(tenantId, span)
+                .compose(tenantDtoResult -> {
+                    if (tenantDtoResult.getStatus() != HttpURLConnection.HTTP_OK) {
+                        TracingHelper.logError(span, "tenant not found");
+                        return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_NOT_FOUND));
+                    }
+                    return Future.succeededFuture(TenantResult.from(
+                            HttpURLConnection.HTTP_OK,
+                            DeviceRegistryUtils
+                                    .convertTenant(tenantId, tenantDtoResult.getPayload(), true),
+                            DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())));
+                });
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final X500Principal subjectDn) {
+        Objects.requireNonNull(subjectDn);
+
+        return get(subjectDn, NoopSpan.INSTANCE);
+    }
+
+    @Override
+    public Future<TenantResult<JsonObject>> get(final X500Principal subjectDn, final Span span) {
+        Objects.requireNonNull(subjectDn);
+        Objects.requireNonNull(span);
+
+        return getForCertificateAuthority(subjectDn, span);
```

Comment:
I don't get the need for having both the `get` and the `getForCertificateAuthority` method.

Reply:
Can be merged. The structure is now obsolete due to refactoring.